### PR TITLE
[CLOSES #1129] DashboardTabProvider

### LINKED ISSUE #1129: Set authentication in AdminDashboardTab bindings for Config and Web-Actions tab from the bound AdminDashboardAccess

### PR DESCRIPTION
**Pull request by adrw**
* New pattern of multibinding tabs to a DashboardTabProvider along with an AccessAnnotation
* Allows for Misk tabs to be set with the authentication from the multibound `AdminDashboardAccess` access annotation
* Also allows for service tabs to be easily set with service specific Access Annotations

* Fixes small bug in Loader tab where dashboard doesn't load when no tabs are bound

### CODE REVIEW

**Comment by adrw** [comment]
Thanks to @mmihic for help with Guice Provider pattern!

File: misk/src/main/kotlin/misk/web/WebTab.kt [hunk 1]
```diff
@@ -10,22 +10,11 @@ abstract class WebTab(
...
-  fun isAuthenticated(caller: MiskCaller?): Boolean {
-    return when {
-      capabilities.isEmpty() && services.isEmpty() -> true // no capabilities/service requirement => unauthenticated requests allowed (including when caller is null)
-      caller == null -> false // capability/service requirement present but caller null => assume authentication broken
-      capabilities.any { caller.capabilities.contains(it) } -> true // matching capability
-      services.any { caller.service == it } -> true // matching service
-      else -> false
-    }
+  fun isAuthenticated(caller: MiskCaller?): Boolean = when {
+    capabilities.isEmpty() && services.isEmpty() -> true // no capabilities/service requirement => unauthenticated requests allowed (including when caller is null)
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/WebTab.kt]: nit for next time: if ya gonna reformat, wrap to 100 please

File: misk/src/main/kotlin/misk/web/metadata/AdminDashboardModule.kt [hunk 2]
```diff
@@ -30,63 +31,66 @@ class AdminDashboardModule(val environment: Environment) : KAbstractModule() {
...
+      slug = "loader",
+      web_proxy_url = "http://localhost:3100/"
     ))
     install(WebTabResourceModule(
-        environment = environment,
-        slug = "loader",
-        web_proxy_url = "http://localhost:3100/",
-        url_path_prefix = "/_admin/",
-        resourcePath = "classpath:/web/_tab/loader/"
+      environment = environment,
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/metadata/AdminDashboardModule.kt]: this reformatting is wrong

File: misk/src/test/kotlin/misk/web/actions/AdminDashboardTabActionTest.kt [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  val module = AdminDashboardActionTestingModule()
+
+  @Inject private lateinit var jetty: JettyService
+  @Inject private lateinit var httpClientFactory: HttpClientFactory
+  @Inject private lateinit var adminDashboardTabAction: AdminDashboardTabAction
+
+  val path = "/api/admindashboardtabs"
+  val empty = "{\"adminDashboardTabs\":[]}"
+  val tabs =
+    "{\"adminDashboardTabs\":[{\"slug\":\"config\",\"url_path_prefix\":\"/_admin/config/\",\"name\":\"Config\",\"category\":\"Container Admin\",\"capabilities\":[\"admin_access\"],\"services\":[]},{\"slug\":\"web-actions\",\"url_path_prefix\":\"/_admin/web-actions/\",\"name\":\"Web Actions\",\"category\":\"Container Admin\",\"capabilities\":[\"admin_access\"],\"services\":[]}]}"
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/web/actions/AdminDashboardTabActionTest.kt]: pretty print this JSON and use a multiline string?

File: misk/src/test/kotlin/misk/web/actions/DashboardTabTest.kt [hunk 4]
```diff
@@ -54,20 +26,23 @@ internal class DashboardTabTest {
 
   @Test
   internal fun tabGoodCategory() {
-    DashboardTab(slug = "slug", url_path_prefix = "/a/path/", name = "Name", category = "@tea-pot_418")
+    DashboardTab(slug = "slug", url_path_prefix = "/a/path/", name = "Name",
+      category = "@tea-pot_418")
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/web/actions/DashboardTabTest.kt]: wrap at +4!

File: misk/src/test/kotlin/misk/web/actions/AdminDashboardTabActionTest.kt (hunk 3 above)
Comment (swankjesse) [reply misk/src/test/kotlin/misk/web/actions/AdminDashboardTabActionTest.kt]: and then you can assert on the object, not on its string representation?
```
val moshi = Moshi.Builder().build()
val responseAdapter = moshi.adapter<AdminDashboardTabAction.Response>()
return responseAdapter.fromJson(response.body!!.source())
```

File: misk/src/main/kotlin/misk/web/metadata/AdminDashboardModule.kt (hunk 2 above)
Comment (adrw) [reply misk/src/main/kotlin/misk/web/metadata/AdminDashboardModule.kt]: Our KtLint configuration changes the `continuation_indent_size` from the Kotlin default 4 to 2:

```
ktlint(dep.ktlintVersion).userData(['indent_size': '2', 'continuation_indent_size' : '2'])
```

File: misk/src/test/kotlin/misk/web/actions/AdminDashboardTabActionTest.kt (hunk 3 above)
Comment (adrw) [reply misk/src/test/kotlin/misk/web/actions/AdminDashboardTabActionTest.kt]: Perfect! I wanted to do this but I didn't know how 👍

File: misk/src/main/kotlin/misk/web/WebTab.kt [hunk 5]
```diff
@@ -10,22 +10,19 @@ abstract class WebTab(
...
 ) : ValidWebEntry(slug = slug, url_path_prefix = url_path_prefix) {
-  fun isAuthenticated(caller: MiskCaller?): Boolean {
-    return when {
-      capabilities.isEmpty() && services.isEmpty() -> true // no capabilities/service requirement => unauthenticated requests allowed (including when caller is null)
-      caller == null -> false // capability/service requirement present but caller null => assume authentication broken
-      capabilities.any { caller.capabilities.contains(it) } -> true // matching capability
-      services.any { caller.service == it } -> true // matching service
-      else -> false
-    }
+  fun isAuthenticated(caller: MiskCaller?): Boolean = when {
```
Comment (wesleyk) [inline misk/src/main/kotlin/misk/web/WebTab.kt]: any particular reason we have to re-implement our authentication logic for web tabs vs web actions? Would be good to have a single source of truth for what is pretty critical security logic

File: misk/src/main/kotlin/misk/web/DashboardTab.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+inline fun <reified A : Annotation> DashboardTabProvider(
+  slug: String,
+  url_path_prefix: String,
+  name: String,
+  category: String = "Admin"
+) = DashboardTabProvider(
+  slug = slug,
+  url_path_prefix = url_path_prefix,
```
Comment (wesleyk) [inline misk/src/main/kotlin/misk/web/DashboardTab.kt]: camelcase?

File: misk/src/main/kotlin/misk/web/DashboardTab.kt [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import javax.inject.Inject
+import javax.inject.Provider
+import kotlin.reflect.KClass
+
+class DashboardTab(
+  slug: String,
+  url_path_prefix: String,
+  val name: String,
+  val category: String = "Admin",
+  capabilities: Set<String> = setOf(),
```
Comment (wesleyk) [inline misk/src/main/kotlin/misk/web/DashboardTab.kt]: do we actually want defaults for this?

File: misk/src/main/kotlin/misk/web/WebTab.kt (hunk 5 above)
Comment (adrw) [reply misk/src/main/kotlin/misk/web/WebTab.kt]: Good point, we can reused the `isAllowed` function from `AccessInterceptor`. Will do so in a follow up PR.

File: misk/src/main/kotlin/misk/web/DashboardTab.kt (hunk 6 above)
Comment (adrw) [reply misk/src/main/kotlin/misk/web/DashboardTab.kt]: Snakecase was recommended in early work on tabs to signify that the variable was exposed as part of the API. I'm open to changing it to camelcase if we decide to set that recommendation aside. @swankjesse

File: misk/src/main/kotlin/misk/web/DashboardTab.kt (hunk 7 above)
Comment (adrw) [reply misk/src/main/kotlin/misk/web/DashboardTab.kt]: I think so, though I'd be open to removing the default and forcing developers to deliberately choose authentication vs. being open by default. Is that what you'd recommend?

File: misk/src/main/kotlin/misk/web/DashboardTab.kt (hunk 6 above)
Comment (wesleyk) [reply misk/src/main/kotlin/misk/web/DashboardTab.kt]: mm gotcha. I guess in my mind that makes sense for the API object itself, which I don't think this provider is

File: misk/src/main/kotlin/misk/web/DashboardTab.kt (hunk 7 above)
Comment (wesleyk) [reply misk/src/main/kotlin/misk/web/DashboardTab.kt]: > deliberately choose authentication vs. being open by default

I think being open by default seems like a dangerous default.

**Pull request merged**